operator/pkg/ciliumendpointbatch/reconciler.go

```diff
@@ -0,0 +1,131 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ciliumendpointbatch
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	cilium_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2alpha1"
+	"github.com/sirupsen/logrus"
+)
+
+// reconciler is used to sync the current (i.e. desired) state of the CEBs in datastore into current state CEBs in the k8s-apiserver.
+// The source of truth is in local datastore, client reconcile with apiserver at regular intervals.
+type reconciler struct {
+	client     clientset.CiliumV2alpha1Interface
+	cebManager cebManager
+}
+
+// newReconciler creates and initializes a new reconciler.
+func newReconciler(client clientset.CiliumV2alpha1Interface, cebMgr cebManager) *reconciler {
+	return &reconciler{
+		client:     client,
+		cebManager: cebMgr,
+	}
+}
+
+// reconcileWithServer gets the list of CEBs to create, update and delete and syncs with the api-server.
+// On error, it returns the list of failed createCebs, updateCebs and deleteCebs to the caller.
+func (r *reconciler) reconcileWithServer(cebsToCreate, cebsToUpdate, cebsToDelete []string) (errCebCreate, errCebUpdate, errCebDelete []string) {
+	var err error
+	var retCeb, ceb *cilium_v2.CiliumEndpointBatch
+
+	// Loop through list of cebsToUpdate and update the CEB
+	for _, c := range cebsToUpdate {
+
+		// Before syncing with ApiServer, get list of removed CEPs
+		remCeps := r.cebManager.getRemovedCeps(c)
+		// Get the copy of Ceb from the cebManager
+		if ceb, err = r.cebManager.getCebFromCache(c); err != nil {
+			continue
+		}
+
+		if retCeb, err = r.client.CiliumEndpointBatches().Update(
+			context.TODO(), ceb, metav1.UpdateOptions{}); err != nil {
+
+			// If CEB is not found in server, instead of resync try re-create new CEB.
+			if errors.IsNotFound(err) {
+				log.WithError(err).WithFields(logrus.Fields{
+					"ceb-name": c,
+				}).Debug("Error from server CEB is notfound in server")
+				// If CEB is not found in server, create the CEB
+				// TODO: Update the local datastore CEB with clean metadata values ???.
+				cebsToCreate = append(cebsToCreate, c)
+			} else {
+				log.WithError(err).WithFields(logrus.Fields{
+					"ceb-name": c,
+				}).Info("Error Updating CiliumEndpointBatch")
+				errCebUpdate = append(errCebUpdate, c)
+			}
+			continue
+		}
+		// Update local datastore CEB with latest object metadata values.
+		r.cebManager.updateCebInCache(retCeb, false)
+		// Since local copy of CEB is synced with api-server, we don't need to track anymore removed
+		// CEPS.
+		r.cebManager.clearRemovedCeps(c, remCeps)
```

Comment:
Now I'm confused as to why we are keeping track of these removed ceps in the first place. Are we not doing anything with them?

Reply:
This goes with the above comment, reason for not deleting the CEPs immediately from the local index is that if one of the CEP is "recreated", we don't want it to end up in another CEB.

Playing devil's advocate, let's say we DO remove CEP from local index immediately then the following chain of event will cause trouble:
- All CEP in CEB X are deleted, including a CEP call "default/my-pod"
- aggregator syncs with apiserver using cebDelete for X
- DELETE for X starts to broadcast to all agents
- CEP "default/my-pod" is recreated and batched into another CEB Y
- aggregator sync with apiserver using cebCreate for Y
- CREATE for Y starts to broadcast to all agents

Now on some remote agent:
- Got CREATE for Y
- creates local copy of CEP "default/my-pod" because it is included in Y
- Got DELETE for X
- deletes local copy of CEP "default/my-pod" because it is included in X
- now we end up in a state where we lost "default/my-pod"

Reply:
I see. But how will we know that the removed CEP still exists in some CEB so as to recreate it in that same CEB again? AFAICT we remove it from the ceptoceb map when we add it to the removedceps field.

Reply:
> AFAICT we remove it from the ceptoceb map when we add it to the removedceps field.

This is what's being changed, we do not remove it immediately from the map but only remove it when the sync goes through with the apiserver.